src/command/shim.rs

```diff
@@ -100,38 +108,41 @@ Options:
     }
 }
 
-fn list(session: &mut Session, verbose: bool) -> Fallible<bool> {
+fn list(session: &Session, verbose: bool) -> Fallible<bool> {
```

Comment:
This is an improvement! But there are still a few things we can do to make it simpler. First a few general points:

- You should generally avoid inspecting error cases, and just propagate them.
- You should generally prefer `let foo = bar()?;` to `bar().map(|foo|, { ... })`. This avoids rightward drift, avoids confusing iteration `map` versus result `map`, and lets you code closer to a style that's like exceptions. As an analogy in JS, this is why `let foo = await bar();` is nicer than `bar.then((foo) => { ... })`.
- It's easiest to just convert foreign error types like `io::Error` into notion's error types as early as possible, and the `?` syntax makes that lightweight. So instead of returning `Result<Foo, io::Result>` from helper functions, just return `Fallible<Foo>`.

As for the `Command` trait, don't throw away errors and return `Ok(false)` -- you want to let the particular error propagate out.

Here's how we could write this function to be simpler and more readable/maintainable:

```rust
fn list(session: &Session, verbose: bool) -> Fallible<()> {
    let shim_dir = path:shim_dir()?;
    let files = fs::read_dir(shim_dir).unknown()?;

    for file in files {
        print_file_info(f, session, verbose)?;
    }

    Ok(())
}
```

Reply:
That is so much simpler!

I made peace with the borrow checker, but lately I have been fighting Result and Option. This helps a lot!